tendermint/src/lite/verifier.rs

```diff
@@ -184,98 +124,120 @@ where
+    trust_threshold: &L,
+    trusting_period: &Duration,
+    now: &SystemTime,
+    store: &mut S,
+) -> Result<(), Error>
+where
+    TS: TrustedState<LastHeader = SH, ValidatorSet = C::ValidatorSet>,
+    SH: SignedHeader<Commit = C>,
+    C: Commit,
+    L: TrustThreshold,
+    S: Store<TrustedState = TS>,
+{
+    // fetch the latest state and ensure it hasn't expired
+    let trusted_state = store.get(Height::from(0))?;
+    let trusted_sh = trusted_state.last_header();
+    is_within_trust_period(trusted_sh.header(), trusting_period, now)?;
 
-    // Can we trust pivot header based on trusted_state?
-    can_trust_bisection(
+    verify_single(
         trusted_state,
-        &pivot_header,
+        untrusted_sh,
+        untrusted_vals,
+        untrusted_next_vals,
         trust_threshold,
-        trusting_period,
-        now,
-        req,
-        store,
     )?;
-    // Trust the header in between the trusted and (still) untrusted height:
-    let pivot_next_vals = req.validator_set(pivot + 1)?;
-    let pivot_trusted = TS::new(&pivot_header, &pivot_next_vals);
-    store.add(&pivot_trusted)?;
-
-    // Can we trust the (still) untrusted header based on the (now trusted) "pivot header"?
-    can_trust_bisection(
-        &pivot_trusted,
-        untrusted_header,
-        trust_threshold,
-        trusting_period,
-        now,
-        req,
-        store,
-    )?;
-    // Add header (and corresponding next validators) to trusted state store:
-    let untrusted_next_vals = req.validator_set(untrusted_header.header().height().increment())?;
-    let untrusted_state = TS::new(untrusted_header, &untrusted_next_vals);
-    store.add(&untrusted_state)?;
 
-    Ok(())
+    // the untrusted header is now trusted. update the store
+    let new_trusted_state = TS::new(untrusted_sh, untrusted_next_vals);
+    store.add(&new_trusted_state)
 }
 
-/// This function captures the high level logic of the light client verification, i.e.,
-/// an application call to the light client module to (optionally download) and
-/// verify a header for some height.
-pub fn verify_header<TS, SH, C, L, S, R>(
-    height: Height,
+/// Attempt to update the store to the given untrusted height
+/// by requesting the necessary data (signed headers and validators).
+/// Returns an error if:
+///     - we're already at or past that height
+///     - our latest state expired
+///     - any requests fail
+///     - requested data is inconsistent (eg. vals don't match hashes in header)
+///     - validators did not correctly commit their blocks
+/// This function is recursive: it uses a bisection algorithm
+/// to request data for intermediate heights as necessary.
+/// Ensures our last trusted header hasn't expired yet, and that
+/// data from the untrusted height can be verified, possibly using
+/// data from intermediate heights.
+/// This function is primarily for use by a light node.
+pub fn verify_and_update_bisection<TS, SH, C, L, R, S>(
```

Comment:
A similiar `and` case. I'd recommend this to be a just `update_bisection` - the error returned will communicate if verification failed.